src/components/LayerSwitcher/AddCustomLayer.tsx

```diff
@@ -0,0 +1,283 @@
+import * as React from 'react';
+import Dialog from '@material-ui/core/Dialog';
+import DialogTitle from '@material-ui/core/DialogTitle';
+import DialogActions from '@material-ui/core/DialogActions';
+import DialogContent from '@material-ui/core/DialogContent';
+import { Button, CircularProgress, TextField } from '@material-ui/core';
+import { loadLayers } from '@dlurak/editor-layer-index';
+import { Autocomplete } from '@material-ui/lab';
+import { useAddLayerContext } from './helpers/AddLayerContext';
+
+type SuccessLayerDataInputProps = {
+  index: any[];
+  onSelect: (layer: any | undefined) => void;
+};
+
+const SuccessLayerInput: React.FC<SuccessLayerDataInputProps> = ({
+  index,
+  onSelect,
+}) => (
+  <Autocomplete
+    options={index.map((l) => l.name)}
+    onChange={(_, val) => {
+      const newLayer = index.find((l) => l.name === val);
+      onSelect(newLayer);
+    }}
+    // eslint-disable-next-line react/jsx-props-no-spreading
+    renderInput={(params) => <TextField {...params} label="Layer" />}
+  />
+);
+
+const LoadingLayerInput = () => (
+  <div
+    style={{
+      display: 'flex',
+      flexDirection: 'column',
+      alignItems: 'center',
+    }}
+  >
+    <CircularProgress />
+    <p
+      style={{
+        fontSize: '1rem',
+        color: 'gray',
+      }}
+    >
+      Fetching the layer index...
+    </p>
+  </div>
+);
+
+const ErrorLayerInput = () => (
+  <div>
+    <p>An error occured while fetching the layer index</p>
+  </div>
+);
+
+const LayerDataInput: React.FC<{ onSelect: (layer: any) => void }> = ({
+  onSelect,
+}) => {
+  const [layerIndex, setLayerIndex] = React.useState([]);
+  const [layerIndexState, setLayerIndexState] =
+    React.useState<'success' | 'loading' | 'error'>('loading');
+
+  React.useEffect(() => {
+    loadLayers()
+      .then((result) => {
+        setLayerIndex(result);
+        setLayerIndexState('success');
+      })
+      .catch(() => {
+        setLayerIndex([]);
+        setLayerIndexState('error');
+      });
+  }, []);
+
+  switch (layerIndexState) {
+    case 'success':
+      return <SuccessLayerInput index={layerIndex} onSelect={onSelect} />;
+    case 'error':
+      return <ErrorLayerInput />;
+    case 'loading':
+      return <LoadingLayerInput />;
+    default:
+      return <LoadingLayerInput />;
+  }
+};
+
+const dynamicPartsRegex = /{((?!y|x|zoom)[a-zA-Z:,]+)}/g;
+
+interface Detailsprops {
+  layer: any;
+  onChange: (values: Record<string, string>) => void;
+  onValidation: (ok: boolean) => void;
+}
+
+const Details: React.FC<Detailsprops> = ({ layer, onChange, onValidation }) => {
```

Comment:
I like your attention to the details such as pretty input title, or aligned image. <3 👍 

![image](https://github.com/zbycz/osmapp/assets/385047/cd2793af-9759-4688-940d-2997df9a79e4)